Added pause/resume/end watch and finalize, with tests.

penntaylor opened the pull request:
This allows a single FileWatcher instance (Ruby API) to be started, paused, resumed, ended, and finalized.

I hesitated about issuing a pull request for this series of commits for the following reasons:
* It adds 46 lines to a codebase of just over 100 lines, so it's by no means a "minor" change. I hate to "bloat up" a small, well-written project.
* It (slightly) complicates the logic in `#watch`
* It contains a complete re-write of `#filesystem_updated?` (Not to worry: spec tests still pass)
* Pause/resume/end are probably atypical use cases. They are very useful in a project I'm currently working on, but unlikely to be high on most people's feature requests.
* The features add roughly 3 seconds to the test suite. (Not horrible, but not great either)

Sounds pretty bad! But I decided to go ahead with a pull request because:
* The extra features may justify the extra code.
* Without `#finalize`, there's no way to ensure client code has a chance to handle existing filesystem changes at application shutdown. IMHO it was begging to be implemented.
* My use of hashed snapshots simplifies the logic in `#filesystem_updated?`. To me anyway.

Even if you decide not to merge the pull request, I thought you'd find the snapshotting and diffing logic interesting, and the finalize method useful. I would also be willing to trim this down by removing pause and resume logic and issuing a new pull request. That would still leave the benefits of snapshots and finalizing, but would remove most of the extra logic from `#watch`, reduce the LOC count by about 20 lines, and shave a couple seconds off the test suite.

thomasfl commented:
This is by far the biggest pull request to filewatcher by far. I am looking forward to testing it out when I get home tomorrow. Just a few quick questions. 

Could the threading be used to replace how the --restart option? Right now the --restart option is totally handled by bin/filewatcher. It does a fork instead of threading.

Couldn't watch_pause() be simplified to just pause(() ? Filewatcher doesn't do much other than watching files.

Could filewatcher be improved evem further by displaying a spinner, by printing /-\|- (like npm does) while it's scanning? Just to give a visual feedback so users now when filewatcher is actually running and how fast it takes to scan the filesystem for changes?

penntaylor commented:
I'll rename #pause_watch, #resume_watch, and #end_watch to simply #pause, #resume, and #end. That makes sense. Should it be #stop instead of #end?

I should have some time tomorrow to look at the restart behavior to see whether threading would simplify that logic or make it more complicated. I should be able to add your spinner idea in at the same time.

thomasfl commented:
Yes, stop() sounds better than end. I am looking forward to see how it will turn out with a spinner. It's easy to implement.

penntaylor commented:
There didn't seem to be any advantage in switching from plain forking to a threaded setup, so I left it as forking but simplified the code.

There aren't any tests for bin/filewatcher. Should some be written?

penntaylor commented:
Well, that test failure only on JRuby-19 is a bit odd.

thomasfl commented:
The jruby 19 failure is indeed strange. I can't recreate the error locally

    $ git clone https://github.com/penntaylor/filewatcher.git
    $ rvm use jruby-d19 --install --binary --fuzzy
    $ bundle exec rake

    bacon -Ilib -Itest --automatic --quiet
    .........
    Finished in 1.084 seconds.

    9 tests, 14 assertions, 0 failures, 0 errors

thomasfl commented:
It's good the simplify bin/filewatcher with Process instead of the homebrewed solution I wrote. There should be some tests for bin/filewatcher too. That means moving functionality to a lib/filewatcher/cli.rb class. I have thought about it myself too.

thomasfl commented:
Tried again last night without success to reproduce the test failure on my mac. It takes time to use Travis to do all the testing, but that seems to be the only way right now. This is the only thing blocking your pull request from being merged. I can add the console spinner myself.

penntaylor commented:
I was also unable to reproduce the jruby issue locally, so I added some print statements to the test to try to understand what was going wrong with Travis. Strangely, the tests then passed on Travis. The print statements are not in a place that should have affected the logic or call sequence in any way. I have no explanation for why the Travis tests magically worked after inserting them.

If you want to check out the spinner I added, be sure to pull from the `end-watch` branch of my fork.

thomasfl commented:
Hi Penn, thanks for the pull-request. I have been busy lately on a new consulting gig and have small kids that takes up time. But I look forward to try out your code some more, and push the new gem in a few days. The new diffing logic is very welcome. I think I have to rewrite the readme too, to make it more compact.

thomasfl commented:
New version have been pushed to rubygems.org. Version number has been bumoed. The default scan interval has been shortened from 1 second to 0.5 seconds, to make the spinner go faster. The credits section has been updated with an extra item.

I think the readme.md file needs an example showing a real world example of how to use start, pause & resume.

 Thanks again for the commit.
